python ecosystem: move from sdist to wheel if possible

haampie opened the pull request:
note: this is auto-generated from https://github.com/spack/pypi-to-spack-package.

move about 1300 packages from `sdist` to `*py3-any-any.whl` with auto-generated run deps, build deps are removed.

todo:

- [x] automate removing `def patch` etc
- [x] get rid of python 2 only wheels `-py2-`.

adamjstewart commented on var/spack/repos/builtin/packages/gptune/package.py:
```diff
@@ -43,7 +43,7 @@ class Gptune(CMakePackage):
     depends_on("py-matplotlib", type=("build", "run"))
     depends_on("py-pyyaml", type=("build", "run"))
-    depends_on("py-scikit-optimize@0.9.0", patches=[patch("space.patch")], type=("build", "run"))
+    depends_on("py-scikit-optimize@0.9.0", type=("build", "run"))
```
Did you confirm that this still builds without this patch? I'm guessing it was there for a reason.

haampie replied:
reverted changes to py-scikit-optimize

adamjstewart commented on var/spack/repos/builtin/packages/py-abipy/package.py:
```diff
@@ -13,52 +13,50 @@ class PyAbipy(PythonPackage):
+    )
 
     variant("gui", default=False, description="Build the GUI")
     variant("ipython", default=False, when="@0.2.0", description="Build IPython support")
```
There variants are no longer being used?

adamjstewart commented on var/spack/repos/builtin/packages/py-adal/package.py:
```diff
@@ -19,10 +19,14 @@ class PyAdal(PythonPackage):
+        depends_on("py-cryptography@1.1:", when="@0.4.3:0.4.5,0.4.7:")
+        depends_on("py-pyjwt@1:", when="@0.4.3:0.4.5,0.4.7:1.2.5")
+        depends_on("py-python-dateutil@2:", when="@0.4.3:0.4.5,0.4.7:1.2.5")
+        depends_on("py-requests@2:", when="@0.4.4:0.4.5,0.4.7:1.2.5")
```
These look very specific, but we only have a single version in this package. Seems like these aren't needed?

adamjstewart commented on var/spack/repos/builtin/packages/py-adb-enhanced/package.py:
```diff
@@ -17,15 +17,19 @@ class PyAdbEnhanced(PythonPackage):
-    version("2.5.10", sha256="9e913d09814ce99974c455a766c5b616a92bca551e657517d6e079882eb19bdb")
-    version("2.5.4", sha256="329ee2e0cfceaa41c591398b365d9acdfd45ffe913c64ac06e1538041986fffb")
-    version("2.5.3", sha256="5a1d5182d1a073b440e862e5481c7a21073eccc3cda7a4774a2aa311fee9bbdc")
-    version("2.5.2", sha256="055676156c1566b8d952b9fdfdd89fc09f2d5d1e3b90b4cdf40858ce9947e2ca")
```
What happened to older versions? No wheels?

haampie replied:
They don't exist https://pypi.org/project/adb-enhanced/#history

adamjstewart commented on var/spack/repos/builtin/packages/py-adb-enhanced/package.py:
```diff
@@ -17,15 +17,19 @@ class PyAdbEnhanced(PythonPackage):
+        depends_on("py-asyncio", when="@:0,2.2:2.5.4")
+        depends_on("py-docopt", when="@:0,1.7.2:")
+        depends_on("py-future", when="@:0,1.8:2.5.4")
+        depends_on("py-psutil", when="@:0,1.9:")
```
Seems like these are much more specific than we need to be

adamjstewart commented on var/spack/repos/builtin/packages/py-agate/package.py:
```diff
@@ -16,13 +16,8 @@ class PyAgate(PythonPackage):
-    depends_on("py-babel@2.0:", type=("build", "run"))
-    depends_on("py-isodate@0.5.4:", type=("build", "run"))
-    depends_on("py-python-slugify@1.2.1:", type=("build", "run"))
-    depends_on("py-leather@0.3.2:", type=("build", "run"))
```
No run-time dependencies?

haampie replied:
Not sure about this one. METADATA file has Requires-dist and has no syntax errors. Just that the info is missing in the PyPI database. 1.6.3 is fine again...

adamjstewart commented on var/spack/repos/builtin/packages/py-aiofiles/package.py:
```diff
@@ -15,11 +15,15 @@ class PyAiofiles(PythonPackage):
+        url="https://pypi.org/packages/f4/2b/078a9771ae4b67e36b0c2a973df845260833a4eb088b81c84b738509b4c4/aiofiles-0.5.0-py3-none-any.whl",
+    )
 
     # Historical dependencies
```
Comments can now be removed

haampie replied:
comments aren't part of `ast.parse` so hard to auto-delete

adamjstewart commented on var/spack/repos/builtin/packages/py-aiofiles/package.py:
```diff
@@ -15,11 +15,15 @@ class PyAiofiles(PythonPackage):
-    version("0.7.0", sha256="a1c4fc9b2ff81568c83e21392a82f344ea9d23da906e4f6a52662764545e19d4")
-    version("0.5.0", sha256="98e6bcfd1b50f97db4980e182ddd509b7cc35909e903a8fe50d8849e02d815af")
-
-    depends_on("python@3.6:3", when="@0.7:", type=("build", "run"))
```
Seems like we lost the upper bound on Python 3

haampie replied:
I'm adding constraints like this: collect a list of versions for which the constraint holds, then simplify to ranges. In this case the marker has `python_version >= "3.6" and python_version < "4"`, which holds for all python versions 3.6.0, ..., 3.13.0 (or whatever the latest is). It gets simplified to a range `@3.6:` w/o upperbound since that's shorter.

adamjstewart replied:
Can we keep the upperbound and simplify `<4` to `<=3`?

haampie replied:
could add a fake python 4.0.0 version, but I don't really see the point to be honest.

adamjstewart replied:
It's only for theoretical future backwards-compatibility. If everything is auto-generated, it isn't hard to update it in the event that Python 4 is ever released...

haampie replied:
I've added a fake Python 4.0.0 so `python@:3` type bounds are added.

adamjstewart replied:
That seems like a significantly worse workaround for an easy to solve issue...

adamjstewart commented on var/spack/repos/builtin/packages/py-amqp/package.py:
```diff
@@ -13,21 +13,41 @@ class PyAmqp(PythonPackage):
+        depends_on("py-vine@1.1.3:", when="@2.2:2.4")
 
     def setup_build_environment(self, env):
         env.set("CELERY_ENABLE_SPEEDUPS", True)
```
Won't be used if we install from wheels

haampie replied:
I'm now automatically removing

```python
def setup_build_environment(): ...
def url_for_version(): ...
def install(): ...
def build_directory(): ...
def patch(): ...
```

plus any function that has `@run_before/after` annotations.

adamjstewart replied:
But wouldn't many of those methods be needed for a reason? For example, to force the build to use a Spack-installed dependency instead of a vendored copy?

adamjstewart commented on var/spack/repos/builtin/packages/py-anndata/package.py:
```diff
@@ -15,16 +15,21 @@ class PyAnndata(PythonPackage):
+        depends_on("py-h5py@3.0.0:", when="@0.8:0.10.5")
+        depends_on("py-importlib-metadata@0.7:", when="@0.7-rc1,0.7:0.8 ^python@:3.7")
+        depends_on(
+            "py-natsort", when="@0.5:0.5.1,0.6.1:0.6.4,0.6.15,0.6.18,0.6.22-rc1,0.7-rc1,0.7:"
```
Good lord. There's only 1 version in the package, is all of this needed?

haampie replied:
Alternative view: why do you/we care?

haampie replied:
OK, I'm truncating old versions in when condition now to `<oldest version in spack> - 5`, or so, then it should be much shorter.

adamjstewart commented on var/spack/repos/builtin/packages/py-appdirs/package.py:
```diff
@@ -15,10 +15,18 @@ class PyAppdirs(PythonPackage):
-    version("1.4.0", sha256="8fc245efb4387a4e3e0ac8ebcc704582df7d72ff6a42a53f5600bbb18fdaadc5")
-
-    patch("setuptools-import.patch", when="@:1.4.0")
-    patch("decode-appdirs.patch", when="@1.4.4")
```
Are we sure these patches are no longer needed?

adamjstewart reviewed (commented):
Only reviewed the first 50 packages or so. Overall impressions:

1. I think it makes most sense to create a new `py` repo alongside `builtin` containing every auto-generated package. Then we can slowly remove `builtin` versions of these as we confirm that they work
2. Before removing `type="build"` everywhere, we still need to fix `spack install --test` so that run-time deps get loaded
3. Packages containing `setup_build_environment` or `patch` likely won't work without them (they were added for a reason), although we'll need to test to be sure
4. It would be cool to see a short-list of packages that changed (more than just build-only deps removed)
5. Some of the dep constraints are overly complicated despite the recipe only containing a single version
6. Most comments can be removed, although this won't be an issue when you generate the entire package from scratch (including docstring, license, homepage, etc.)
7. We should probably skip packages with maintainers in the first round, we can port those in the second round

Overall, great to see the Python ecosystem finally being wrangled!

haampie committed:
python ecosystem: sdist -> universal wheels

haampie committed:
undo: py-geopmdpy

haampie committed:
fix ordering

haampie committed:
manual changes

haampie commented:
High-level diff:

```
py-a2wsgi@1.6.0:1.7.0
  removed: py-pdm-pep517
py-acme-tiny
  removed: master
py-acme-tiny@4.0.4
  removed: py-setuptools
py-adb-enhanced
  removed: 2.5.2,2.5.3
py-adb-enhanced@2.5.4:2.5.10
  removed: py-setuptools
py-agate-dbf@0.2.1
  removed: py-dbfread, py-agate
py-agate@1.6.1
  removed: py-isodate, py-pytimeparse, py-parsedatetime, py-python-slugify, py-six, py-babel, py-leather
py-alive-progress@1.6.2:2.4.1
    allow: ^python@3.9.0:3.12.1
py-altair@5.1.0:5.2.0
     drop: ^python@3.7.0:3.7.17
py-amqp@2.4.1:2.4.2
    allow: ^py-vine@5.0.0
py-ansiwrap@0.8.4
  removed: py-textwrap3
py-apispec@4.7.1
  removed: py-packaging
py-argcomplete@1.12.0:3.0.8
  removed: py-importlib-metadata
py-argparse-dataclass@2.0.0
     drop: ^python@3.7.0:3.7.17
py-arrow@0.14.1
  removed: py-python-dateutil
py-arrow@0.16.0
     drop: ^py-python-dateutil@1.5:2.5.2
py-astpretty@2.0.0
  removed: py-setuptools
py-astroid
  removed: 2.2.0,2.0.4,2.2.5,2.3.3
py-astroid@1.4.1:1.4.5
  removed: py-setuptools, py-lazy-object-proxy, py-six, py-wrapt
py-astroid@1.6.6
  removed: py-setuptools
py-astroid@2.4.2
  removed: py-setuptools
    allow: ^py-wrapt@1.13.3:1.15.0
py-astroid@2.5.6
  removed: py-setuptools, py-six
    allow: ^py-lazy-object-proxy@1.6.0:1.7.1
py-astroid@2.12.7:2.14.2
  removed: py-setuptools
    allow: ^python@3.7.0:3.7.1
py-astroid
  removed: 2.2.0,2.0.4,2.2.5,2.3.3
py-astroid@1.4.1:1.4.5
  removed: py-setuptools, py-lazy-object-proxy, py-six, py-wrapt
py-astroid@1.6.6
  removed: py-setuptools
py-astroid@2.4.2
  removed: py-setuptools
    allow: ^py-wrapt@1.13.3:1.15.0
py-astroid@2.5.6
  removed: py-setuptools, py-six
    allow: ^py-lazy-object-proxy@1.6.0:1.7.1
py-astroid@2.12.7:2.14.2
  removed: py-setuptools
    allow: ^python@3.7.0:3.7.1
py-async-timeout@4.0.0
    added: py-typing-extensions
py-autocfg@0.0.8
    added: py-dataclasses
py-auxlib@0.0.43
  removed: py-wheel
py-ax-platform@0.3.1
     drop: ^py-botorch@0.8.4
py-azure-cli-core@2.9.1
    added: py-azure-cli-nspkg
    allow: ^py-pyopenssl@20.0.1
py-azure-cli-telemetry@1.0.4
    added: py-azure-cli-nspkg
py-azure-cli@2.9.1
     drop: ^py-azure-mgmt-cdn@4.1.0rc1
    allow: ^py-azure-mgmt-cdn@4.1.0-rc1
     drop: ^py-azure-mgmt-containerregistry@3.0.0rc14
    allow: ^py-azure-mgmt-containerregistry@3.0.0-rc14
     drop: ^py-azure-mgmt-eventgrid@3.0.0rc7
    allow: ^py-azure-mgmt-eventgrid@3.0.0-rc7
     drop: ^py-azure-mgmt-redis@7.0.0rc1
    allow: ^py-azure-mgmt-redis@7.0.0-rc1
py-azure-identity@1.13.0
    added: py-six
py-azure-keyvault@1.1.0
    added: py-azure-nspkg
py-azure-loganalytics@0.1.0
    added: py-azure-nspkg
py-azure-mgmt-applicationinsights@0.1.1
    added: py-azure-mgmt-nspkg
py-azure-mgmt-batch@17.1.0
    added: py-typing-extensions
py-azure-mgmt-batchai@2.0.0
    added: py-azure-mgmt-nspkg
py-azure-mgmt-billing@0.2.0
    added: py-azure-mgmt-nspkg
py-azure-mgmt-cdn
  removed: 4.1.0rc1
    added: 4.1.0-rc1
py-azure-mgmt-consumption@2.0.0
    added: py-azure-mgmt-nspkg
py-azure-mgmt-containerregistry
  removed: 3.0.0rc14
    added: 3.0.0-rc14
py-azure-mgmt-datamigration@0.1.0
    added: py-azure-mgmt-nspkg
py-azure-mgmt-dns@2.1.0
    added: py-azure-mgmt-nspkg
py-azure-mgmt-eventgrid
  removed: 3.0.0rc7
    added: 3.0.0-rc7
py-azure-mgmt-iothubprovisioningservices@0.2.0
    added: py-azure-mgmt-nspkg
py-azure-mgmt-maps@0.1.0
    added: py-azure-mgmt-nspkg
py-azure-mgmt-msi@0.2.0
    added: py-azure-mgmt-nspkg
py-azure-mgmt-redis
  removed: 7.0.0rc1
    added: 7.0.0-rc1
py-azure-mgmt-relay@0.1.0
    added: py-azure-mgmt-nspkg
py-azure-storage-nspkg@3.1.0
  removed: py-azure-nspkg
py-azureml-automl-core@1.23.0
     drop: ^python@3.9.0:3.12.1
py-azureml-core@1.8.0:1.11.0
     drop: ^py-azure-mgmt-containerregistry@3.0.0rc14
    allow: ^py-azure-mgmt-containerregistry@3.0.0-rc14
    allow: ^py-cryptography@3.3.2:41.0.3
    allow: ^py-pyopenssl@20.0.1:23.2.0
     drop: ^py-ruamel-yaml@0.16.5
py-azureml-core@1.23.0
     drop: ^py-azure-mgmt-containerregistry@3.0.0rc14
    allow: ^py-azure-mgmt-containerregistry@3.0.0-rc14
     drop: ^py-cryptography@3.2.1
    allow: ^py-pyopenssl@20.0.1
py-azureml-dataset-runtime@1.23.0
     drop: ^py-numpy@1.19.4:main
py-azureml-sdk@1.23.0
    allow: ^python@3.9.0:3.12.1
py-azureml-train-automl-client@1.23.0
     drop: ^python@3.9.0:3.12.1
py-babel@2.3.4:2.6.0
  removed: py-setuptools
    allow: ^py-pytz@2014.9:2015.4
py-babel@2.7.0:2.12.1
  removed: py-setuptools
py-backports-tempfile@1.0
  removed: py-backports-weakref
py-backports-weakref
  removed: 1.0rc1
    added: 1.0-rc1
py-bakta@1.5.1
  removed: amrfinder, blast-plus, py-deepsig-biocomp, infernal, aragorn, prodigal, py-setuptools, diamond, trnascan-se, hmmer, pilercr
py-bandit@1.7.0
  removed: py-gitpython, py-colorama, py-six, py-stevedore, py-pyyaml
py-barectf@3.1.2
    added: py-jsonschema, py-termcolor, py-jinja2, py-setuptools, py-pyyaml
py-beautifulsoup4@4.4.1:4.5.3
  removed: py-lxml, py-html5lib
py-bids-validator@1.13.1
    allow: ^python@3.7.0:3.7.17
py-bidscoin@3.7.4:4.1.1
  removed: dcm2niix
py-bidskit@2022.10.13:2023.9.7
  removed: dcm2niix
py-binaryornot@0.4.4
  removed: py-chardet
py-biobb-gromacs@4.1.1
  removed: gromacs
py-biopandas@0.2.5
    added: py-setuptools
py-blight@0.0.47
    allow: ^py-flake8@2.6.2
    allow: ^py-isort@4.2.15:4.3.20
py-bluepyefe@2.3.6
     drop: ^python@3.7.0:3.7.17
py-bluepyopt@1.14.4
  removed: neuron
py-bluepyopt@1.14.4
  removed: neuron
py-boto3@1.9.169:1.10.50
  removed: py-jmespath, py-s3transfer, py-botocore
py-boto3@1.17.112
     drop: ^py-botocore@1.20.27
     drop: ^py-s3transfer@0.3.4
    allow: ^py-s3transfer@0.4.2
py-boto3@1.18.12
    allow: ^py-botocore@1.21.12
py-boto3@1.34.44
     drop: ^python@3.7.0:3.7.17
py-botocore@1.12.169:1.13.50
  removed: py-urllib3, py-docutils, py-jmespath, py-python-dateutil
py-botocore@1.34.44
     drop: ^python@3.7.0:3.7.17
py-botorch@0.8.3
     drop: ^py-gpytorch@1.10
     drop: ^py-linear-operator@0.4.0
py-breathe@4.7.0:4.7.3
  removed: doxygen
     drop: ^py-sphinx@1.3.1
    allow: ^py-sphinx@1.4,4.0.0:7.2.6
py-breathe@4.8.0:4.11.1
  removed: py-sphinx, py-docutils, doxygen, py-six
py-breathe@4.21.0
  removed: py-sphinx, py-docutils, py-pygments, doxygen, py-six
py-breathe@4.33.1:4.35.0
  removed: py-pygments, doxygen
py-build@0.7.0:1.0.3
  removed: py-colorama
py-catalogue@1.0.0:2.0.0
  removed: py-importlib-metadata
py-catkin-pkg@0.4.23
  removed: py-setuptools
py-celery
  removed: 5.0.0,4.4.7,4.3.0
py-certipy@0.1.3
    allow: ^py-pyopenssl@20.0.1
py-cf-xarray@0.9.0
    allow: ^py-xarray@0.9.1:0.18.2
py-cfgrib@0.9.8.5
  removed: py-eccodes
py-checkm-genome@1.2.1
  removed: pplacer, hmmer, prodigal
py-cherrypy@18.1.1
    added: py-pywin32
py-chex@0.1.0:0.1.5
    added: py-dataclasses
py-cleo
  removed: 1.0.0a5
    added: 1.0.0-alpha5
py-click-option-group@0.3.0:0.5.2
  removed: py-setuptools
     drop: ^py-click@8.0.1:8.1.7
py-click-option-group@0.5.3:0.5.6
  removed: py-setuptools
py-click-repl@0.1.6:0.2.0
  removed: py-click, py-prompt-toolkit, py-six
py-climax
  removed: main
py-climax@0.3.1
  removed: py-click, py-xesmf, py-tqdm, py-numpy, py-timm, py-torch, py-pytorch-lightning, py-xarray, py-scipy, py-torchdata
py-cma@3.3.0
  removed: py-moarchiving
py-cmake-format@0.6.9:0.6.11
  removed: py-jinja2
py-cmocean@2.0
  removed: py-numpy, py-matplotlib
py-cmseq@1.0.4
  removed: samtools
    allow: ^py-biopython@1.78:1.81
py-cmyt
  removed: main
py-codespell@2.2.6
     drop: ^python@3.7.0:3.7.17
py-coloredlogs@10.0
    added: py-colorama
py-coloredlogs@14.0
     drop: ^py-humanfriendly@4.18
py-colorful@0.5.4
    added: py-colorama
py-colorlog@3.1.4:4.0.2
    added: py-colorama
py-confection@0.0.4
    allow: ^py-pydantic@1.8.2
py-contextily@1.0.1
    added: py-pillow
py-copulas
    added: 0.6.1
py-copulas@0.6.0
     drop: ^py-numpy@1.20.0:1.26.4
    allow: ^py-scipy@main
     drop: ^python@3.10.0:3.12.1
py-coveralls@3.0.1
  removed: py-pyyaml
    added: py-requests
py-cssselect
  removed: 1.0.2
py-ctgan
    added: 0.5.2
py-ctgan@0.5.0
     drop: ^py-numpy@1.20.0:1.26.4
    allow: ^py-rdt@0.6.4
    allow: ^py-scikit-learn@master
    allow: ^py-torchvision@main
     drop: ^python@3.10.0:3.12.1
py-currentscape@1.0.12
  removed: py-setuptools
py-cwl-upgrader@1.2.4
    added: py-setuptools
py-cwltool@3.1.20211107152837
  removed: py-cwl-utils
    added: py-setuptools
     drop: ^py-rdflib@6.2.0
py-cwltool@3.1.20221109155812:3.1.20221201130942
    added: py-setuptools
py-cylc-flow@8.1.4
  removed: graphviz
     drop: ^py-metomi-isodatetime@3.0.0
    allow: ^py-metomi-isodatetime@1.3.0.0
     drop: ^py-setuptools@68.0.0:69.2.0
py-cylc-flow@8.2.0
  removed: graphviz
     drop: ^py-metomi-isodatetime@3.0.0
    allow: ^py-metomi-isodatetime@1.3.0.0
py-cylc-flow@8.2.3
  removed: graphviz
     drop: ^py-metomi-isodatetime@3.0.0,3.1.0
    allow: ^py-metomi-isodatetime@1.3.0.0:1.3.1.0
py-cylc-rose@1.3.0
     drop: ^py-metomi-isodatetime@3.0.0,3.1.0
    allow: ^py-metomi-isodatetime@1.3.0.0:1.3.1.0
py-cylc-uiserver@1.3.0
    allow: ^py-tornado@6.1
py-dasbus@1.7
  removed: py-pygobject
py-dask-mpi@2.0.0:2.21.0
  removed: py-mpi4py, py-dask
py-dask-mpi@2022.4.0
  removed: py-mpi4py, py-dask, py-distributed
py-dask-mpi/remove-dependency-jupyter@2.0.0,2.21.0
  removed: py-mpi4py, py-dask
py-dask-mpi/remove-dependency-jupyter@2022.4.0
  removed: py-mpi4py, py-dask, py-distributed
py-datacube@1.8.3
  removed: py-setuptools, py-lark
    added: py-lark-parser
py-datalad-container@1.1.5:1.1.7
     drop: ^py-datalad@0.17.5
py-datalad-deprecated@0.3.0
    added: py-pillow
py-datalad-hirni@0.0.8
  removed: git-annex
     drop: ^py-datalad@0.17.5
py-datalad-metalad@0.2.1
  removed: git-annex, py-datalad-metadata-model
     drop: ^py-datalad@0.17.5
py-datalad-metalad@0.4.5
  removed: git-annex
     drop: ^py-datalad@0.17.5
py-datalad-metalad@0.4.17
  removed: git-annex
py-datalad-neuroimaging@0.3.1:0.3.3
  removed: git-annex
     drop: ^py-datalad@0.17.5
py-datalad-webapp@0.3
     drop: ^py-datalad@0.17.5
py-datalad
  removed: 0.17.5
py-datalad@0.14.6:0.15.5
  removed: git, git-annex
    added: py-pillow
    allow: ^py-chardet@5.0.0:5.2.0
py-datalad@0.18.3:0.18.4
  removed: py-exifread, git, py-mutagen, py-python-xmp-toolkit, git-annex, py-pyyaml
py-datasets@1.8.0
    added: py-dataclasses
    allow: ^py-fsspec@0.9.0:2024.2.0
py-dateparser@0.7.2
  removed: py-jdatetime, py-convertdate, py-ruamel-yaml, py-umalqurra
    allow: ^py-python-dateutil@1.5:2.5.2
py-dcm2bids@2.1.9
  removed: dcm2niix
py-dcm2bids@3.1.0
  removed: dcm2niix
    allow: ^python@3.7.0:3.7.17
py-debtcollector@2.2.0
    added: py-pbr
py-deepecho@0.3.0.post1
     drop: ^py-numpy@1.20.0:1.26.4
py-deephyper
  removed: master
py-devito@4.8.1
  removed: intel-parallel-studio, py-matplotlib, py-pandas, intel-oneapi-compilers
    added: py-pip
    allow: ^py-flake8@2.6.2
    allow: ^py-mpi4py@master
py-dh-scikit-optimize@0.9.0:0.9.5
  removed: py-configspace
py-dictdiffer@0.8.1
  removed: py-setuptools
py-django@2.2.10:2.2.12
  removed: py-asgiref
py-django@3.0.1:3.0.5
     drop: ^py-asgiref@3.1.3:3.1.4
py-domdf-python-tools@3.6.1
    added: py-importlib-resources
py-downhill@0.4.0
  removed: py-click, py-theano
py-ecdsa@0.13.2
  removed: py-six
py-elastic-transport@8.4.0
    added: py-dataclasses
py-elasticsearch@2.3.0:7.5.1
  removed: py-urllib3, py-aiohttp
py-elasticsearch@7.6.0
  removed: py-aiohttp
py-exhale@0.3.6
    added: py-sphinx, py-docutils, py-breathe
py-exodus-bundler@2.0.2
  removed: py-setuptools, musl
py-fabric3@1.14.post1
  removed: py-paramiko, py-six
py-faker
    added: 19.13.0
py-fastai@2.5.3
    added: py-packaging, py-pip, py-pillow
py-fastapi@0.88.0:0.98.0
     drop: ^py-ujson@4.0.2
py-fastcore@1.3.27
    added: py-packaging, py-pip
py-fava@1.18
  removed: py-setuptools, py-pytest
py-file-magic@0.4.1
  removed: file
py-flake8-polyfill@1.0.2
    allow: ^py-flake8@2.6.2
py-flake8
    added: 2.6.2
py-flake8@2.5.4
  removed: py-pycodestyle, py-setuptools
    added: py-pep8
py-flake8@3.0.4:5.0.4
  removed: py-setuptools
py-flake8@6.0.0:6.1.0
  removed: py-setuptools, py-importlib-metadata
    allow: ^python@3.8.0
py-flask-restful@0.3.9
  removed: py-pytz, py-aniso8601, py-six, py-flask
py-flask-socketio@2.9.6
  removed: py-werkzeug, py-python-socketio, py-flask
py-flask@0.11.1:0.12.2
  removed: py-click, py-jinja2, py-itsdangerous, py-werkzeug, py-setuptools
py-flask@0.12.4
  removed: py-setuptools
    allow: ^py-jinja2@2.7:2.10
    allow: ^py-werkzeug@0.11.11:0.12.2
py-flask@1.1.1:2.2.2
  removed: py-setuptools
py-flask@2.3.2
  removed: py-setuptools
     drop: ^py-click@8.0.1:8.0.3
     drop: ^py-itsdangerous@2.0.1
     drop: ^py-jinja2@3.0.1:3.0.3
py-flax@0.8.1
     drop: ^py-numpy@1.22.0:1.25.2
py-flexx@0.4.1
  removed: py-tornado
py-fonttools
  removed: 4.39.4
py-fracridge@1.4.3:2.0
    added: py-pillow
    allow: ^py-setuptools-scm@1.15.6:3.3.3
py-freezegun@0.3.12
  removed: py-six, py-python-dateutil
py-furo@2023.5.20
     drop: ^py-sphinx-basic-ng@1.0.0b2
    allow: ^py-sphinx-basic-ng@1.0.0-beta2
py-furo@2023.9.10
     drop: ^py-sphinx-basic-ng@1.0.0b2
    allow: ^py-sphinx-basic-ng@1.0.0-beta2
     drop: ^python@3.7.0:3.7.17
py-galaxy-sequence-utils@1.1.5
    added: py-six
py-gcovr@4.2:5.2
  removed: py-setuptools
py-gcsfs@2024.2.0
     drop: ^python@3.7.0:3.7.17
py-gdbgui@0.13.2.0
  removed: py-setuptools, gdb
py-geeadd@0.3.0
    added: py-pytest
py-genshi@0.7.7
  removed: py-setuptools
py-geoalchemy2@0.4.2:0.6.3
  removed: py-shapely
py-geomdl@5.3.1
  removed: py-numpy, py-plotly, py-matplotlib
py-geopandas
  removed: master
py-geopandas@0.3.0
    added: py-descartes
    allow: ^py-shapely@2.0.0:main
py-geopandas@0.4.0,0.8.1:0.10.2,0.14.3
    allow: ^py-shapely@2.0.0:main
py-geopandas@0.5.0
  removed: py-pyproj, py-pandas, py-fiona, py-shapely
py-geoplot@0.4.1
     drop: ^py-geopandas@master
py-gidgethub
  removed: main
py-gidgetlab
  removed: main
py-gidgetlab@1.1.0
  removed: py-setuptools
py-gimmik@2.2:2.3
    allow: ^python@3.7.0:3.7.17,3.9.0:3.12.1
py-gimmik@3.0
    allow: ^python@3.7.0:3.8.18
py-gitdb@4.0.6
     drop: ^py-smmap@4.0.0
py-globus-sdk@3.7.0:3.10.1
    added: py-typing-extensions
py-glymur@0.9.9
    added: py-setuptools
py-gooey@1.0.8.1
     drop: ^py-wxpython@4.0.6
py-google-api-core@2.11.0:2.17.0
  removed: py-setuptools
py-google-api-python-client@1.7.10
     drop: ^py-uritemplate@4.1.1
py-google-auth@1.6.3
  removed: py-setuptools, py-aiohttp
py-google-auth@2.3.2
     drop: ^py-setuptools@20.6.7:40.2.0
py-google-auth@2.11.0:2.27.0
  removed: py-setuptools
py-google-cloud-appengine-logging@1.4.1
     drop: ^py-protobuf@3.20.0:3.20.1,4.21.5
py-google-cloud-audit-log@0.2.5
     drop: ^py-protobuf@3.20.0:3.20.1,4.21.5
py-google-cloud-batch@0.17.11
     drop: ^py-protobuf@3.20.0:3.20.1,4.21.5
py-google-cloud-logging@3.9.0
     drop: ^py-protobuf@3.20.0:3.20.1,4.21.5
py-google-pasta@0.1:0.1.7
  removed: py-six
py-gql
  removed: 3.0.0a1
    added: 3.0.0-alpha1
py-gql@0.4.0
  removed: py-six, py-promise, py-graphql-core, py-requests
py-gradio@3.36.1
    added: py-pillow
py-grandalf@0.6
    added: py-future
py-graphene-tornado@2.6.1
     drop: ^py-werkzeug@0.15.3
    allow: ^py-werkzeug@0.12.2
py-graphviz@0.8.4
    allow: ^py-flake8@2.6.2
    allow: ^py-sphinx@1.3.1:1.6.3
py-graphviz@0.10.1:0.13.2
    allow: ^py-flake8@2.6.2
py-grayskull@2.5.0
    allow: ^py-setuptools@39.0.1:59.4.0
py-grpc-google-iam-v1@0.13.0
     drop: ^py-protobuf@3.20.0:3.20.1,4.21.5
py-gtdbtk@2.1.0,2.3.2
  removed: pplacer, prodigal, mash, fastani, hmmer, fasttree
py-gtdbtk@2.3.0
  removed: pplacer, prodigal, mash, fastani, hmmer, fasttree
    allow: ^py-pydantic@1.8.2
py-gxformat2@0.16.0
  removed: py-schema-salad
py-h11@0.13.0
    added: py-dataclasses
py-h5sh@0.1.1
  removed: py-setuptools
py-hacking@1.1.0
    added: py-pbr, py-flake8, py-six
py-hatchling@1.4.1:1.17.0
    added: py-importlib-metadata
py-hatchling@1.18.0:1.21.0
     drop: ^python@3.7.0:3.7.17
py-heat@1.3.0
    added: py-pillow
py-hepunits
  removed: master
py-hieroglyph@1.0.0
    allow: ^py-sphinx@1.4
py-histoprint@2.2.0:2.4.0
    allow: ^py-click@7.0
py-httpbin@0.5.0
  removed: py-brotlipy, py-raven, py-werkzeug
py-httpx@0.22.0
    added: py-async-generator
py-huggingface-hub@0.0.8:0.0.10,0.14.1
    added: py-importlib-metadata
py-huggingface-hub@0.10.1
    added: py-inquirerpy, py-importlib-metadata
py-huggingface-hub@0.19.4
     drop: ^python@3.7.0:3.7.17
py-hyperlink@21.0.0
  removed: py-idna
py-hyperopt@0.2.5
  removed: py-pymongo, py-pyspark
py-idna@2.5:3.3
  removed: py-setuptools
py-imagehash@4.3.1
    added: py-pillow
py-imbalanced-learn@0.10.1
    allow: ^python@3.7.0:3.7.17
py-ipycanvas@0.9.0:0.10.2
    added: py-pillow
py-ipykernel@4.1.0:4.2.1
  removed: py-tornado, py-jupyter-client, py-traitlets, py-ipython
py-ipykernel@4.2.2
  removed: py-tornado
    allow: ^py-traitlets@4.0.0
py-ipykernel@4.3.0:4.3.1
    allow: ^py-traitlets@4.0.0
py-ipympl@0.8.8
    added: py-pillow
py-ipython@5.1.0
    added: py-win-unicode-console
     drop: ^py-simplegeneric@0.8
py-ipython@5.8.0
     drop: ^py-simplegeneric@0.8
py-ipython@7.3.0
    added: py-jedi
py-ipython@7.18.1
    allow: ^py-jedi@0.10.0:0.15.1
    allow: ^py-prompt-toolkit@2.0.9:2.0.10
py-ipython@7.21.0
    allow: ^py-prompt-toolkit@2.0.9:2.0.10
py-ipython@8.0.1
    added: py-setuptools
py-ipywidgets@5.2.2
  removed: py-traitlets, py-ipykernel, py-ipython, py-jupyterlab-widgets, py-widgetsnbextension
py-ipywidgets@7.4.2:7.5.1
  removed: py-jupyterlab-widgets
py-isa-rwval@0.10.10
     drop: ^py-networkx@2.5.1
py-isodate@0.6.0
  removed: py-six
py-isort@4.2.15:4.3.20
  removed: py-setuptools, py-colorama
py-iterative-stats
  removed: main
py-jedi@0.9.0:0.18.2
  removed: py-setuptools
py-jinja2-time@0.2.0
  removed: py-jinja2, py-arrow
py-jmp@0.0.2
  removed: py-jax
    added: py-dataclasses
py-joblib@0.10.0:1.2.0
  removed: py-setuptools
py-jsonpatch@1.23
    allow: ^py-jsonpointer@2.0
py-jupyter-console@4.0.2:4.0.3
  removed: py-prompt-toolkit, py-jupyter-client, py-ipykernel, py-ipython, py-pygments
py-jupyter-console@4.1.0:4.1.1
  removed: py-prompt-toolkit, py-pygments
    added: py-pyreadline
    allow: ^py-ipython@7.3.0:8.14.0
py-jupyter-console@5.0.0:5.2.0
    allow: ^py-ipython@7.3.0:8.14.0
py-jupyter-core@4.0.0:4.0.6
  removed: py-setuptools, py-traitlets
py-jupyter-core@4.1.0:4.4.0
  removed: py-setuptools
py-jupyter-core@4.6.0:4.9.2
  removed: py-setuptools
    added: py-pywin32
py-jupyter-core@4.11.1:5.3.0
    added: py-pywin32
py-jupyter-packaging@0.7.12
  removed: py-setuptools
py-jupyter-packaging@0.10.4:0.11.1
    allow: ^py-setuptools@62.3.2:69.2.0
py-jupyter-server@1.6.1
  removed: py-websocket-client, py-requests-unixsocket
    added: py-pywin32
    allow: ^py-anyio@4.0.0
    allow: ^py-jupyter-core@4.4.0
py-jupyter-server@1.6.1
  removed: py-websocket-client, py-requests-unixsocket
    added: py-pywin32
    allow: ^py-anyio@4.0.0
    allow: ^py-jupyter-core@4.4.0
py-jupyter@1.0.0
  removed: py-notebook, py-ipykernel, py-qtconsole, py-nbconvert, py-jupyter-console, py-ipywidgets
py-jupyterhub@0.9.4:1.0.0
  removed: py-notebook, py-mako, py-setuptools
py-jupyterhub@1.4.1
  removed: py-notebook, py-mako, py-setuptools
    added: py-psutil
py-jupyterlab-pygments@0.2.2
  removed: py-pygments
py-jupyterlab@2.1.0
  removed: py-setuptools
    allow: ^py-jupyterlab-server@1.1.0,2.6.0:2.22.1
    allow: ^python@3.10.0:3.12.1
py-jupyterlab@2.2.7
  removed: py-setuptools
py-jupyterlab@3.0.16:3.4.8
  removed: py-setuptools
    allow: ^py-jinja2@2.7:2.9.6
py-jupyterlab@4.0.1
  removed: py-hatch-jupyter-builder, py-hatchling
py-jupytext@1.13.0:1.14.1
  removed: py-jupyterlab
py-jupytext@1.16.0
  removed: py-jupyterlab, py-hatch-jupyter-builder
py-keras-applications
  removed: 1.0.3
py-keras-applications@1.0.1:1.0.4
    added: py-h5py, py-numpy, py-keras
py-keras-applications@1.0.6:1.0.8
    added: py-h5py, py-numpy
py-keras-preprocessing@1.0.1:1.0.3
    added: py-scipy, py-keras
py-keystoneauth1@4.3.1
    added: py-pbr
py-kombu@4.3.0:4.5.0
  removed: py-cached-property, py-importlib-metadata
    allow: ^py-amqp@2.4.1:2.4.2,2.6.1
    allow: ^py-redis@3.3.8
py-kombu@4.6.6:5.0.2
  removed: py-cached-property
py-kornia@0.7.0:0.7.2
     drop: ^python@3.7.0:3.7.17
py-kosh@3.0.1
    allow: ^py-hdbscan@0.8.26
py-kubernetes@9.0.0:10.0.1,11.0.0:26.1.0
    added: py-setuptools
py-kubernetes@10.1.0
    added: py-setuptools
     drop: ^py-pyyaml@5.1:6.0
py-langsmith@0.0.7:0.0.11
    allow: ^python@3.8.0
py-latexcodec@1.0.4
  removed: py-six
py-leather@0.3.3
  removed: py-six
py-lightly-utils@0.0.2
    added: py-pillow
py-lightly
  removed: 1.4.9
py-lightly@1.4.1
  removed: py-torch
     drop: ^py-pytorch-lightning@1.6.0,1.6.1,1.6.2,1.6.3,1.6.4,1.6.5,1.7.0,1.7.1,1.7.2,1.7.3,1.7.4,1.7.5,1.7.6,1.7.7,1.8.0,1.8.1,1.8.2,1.8.3
    allow: ^py-pytorch-lightning@2.0.0
py-lightly@1.4.2:1.5.0
     drop: ^py-pytorch-lightning@1.6.0,1.6.1,1.6.2,1.6.3,1.6.4,1.6.5,1.7.0,1.7.1,1.7.2,1.7.3,1.7.4,1.7.5,1.7.6,1.7.7,1.8.0,1.8.1,1.8.2,1.8.3
py-lightning-utilities@0.3.0:0.5.0,0.8.0
  removed: py-setuptools
py-lightning-utilities@0.6.0.post0
  removed: py-setuptools
     drop: ^py-packaging@17.1:19.2
py-llnl-sina@1.10.0:1.11.0
  removed: py-ujson
py-llnl-sina@1.10.0:1.11.0
  removed: py-ujson
py-lmodule@0.1.0
  removed: lmod
py-macholib@1.11:1.16
  removed: py-setuptools
py-markdown-it-py@1.1.0,3.0.0
    added: py-mdit-py-plugins
py-markdown-it-py@2.2.0
    added: py-mdit-py-plugins
    allow: ^python@3.7.0:3.7.17
py-markdown@3.1.1
    added: py-setuptools
py-matplotlib-inline@0.1.2:0.1.6
  removed: py-matplotlib
py-matplotlib-scalebar
  removed: master
py-mdit-py-plugins@0.3.1
  removed: py-flit-core
py-merlin
  removed: develop,master
py-merlin@1.10.3
  removed: py-setuptools
    added: py-importlib-resources
py-metaphlan@3.1.0
  removed: blast-plus, samtools, muscle, raxml, py-matplotlib, bowtie2
py-metaphlan@4.0.2
  removed: blast-plus, py-matplotlib, samtools, muscle, raxml, bowtie2
py-metomi-isodatetime
  removed: 3.0.0,3.1.0
    added: 1.3.1.0,1.3.0.0
py-metomi-rose@2.1.0
     drop: ^py-metomi-isodatetime@3.0.0,3.1.0
    allow: ^py-metomi-isodatetime@1.3.0.0:1.3.1.0
py-mizani@0.8.1
    added: py-tzdata
py-mlflow@1.17.0
  removed: py-pyarrow, py-jinja2, py-importlib-metadata, py-markdown, py-packaging, py-scipy, py-matplotlib, py-scikit-learn, py-shap
    added: py-prometheus-flask-exporter
     drop: ^py-alembic@1.5.5
    allow: ^py-numpy@main
    allow: ^py-pandas@2.0.0:2.2.1
    allow: ^py-protobuf@3.6.0:3.11.2
    allow: ^py-pytz@2023.3
    allow: ^py-pyyaml@3.11:3.13
    allow: ^py-sqlalchemy@1.0.12:1.3.19,2.0.19
    allow: ^py-sqlparse@0.3.1
py-mlflow@2.0.1
    allow: ^py-scikit-learn@master
    allow: ^py-scipy@main
py-mlxtend@0.16.0
    added: py-setuptools
py-moarchiving@0.6.0
  removed: py-fraction
py-mock@1.3.0:2.0.0
  removed: py-six
py-mock@3.0.3:3.0.5
    added: py-six
py-modin@0.16.2
  removed: py-ray, py-pyarrow, py-redis, py-dask, py-distributed, py-pickle5
    allow: ^py-pandas@1.5.1
py-modules-gui@0.2
  removed: environment-modules
py-motmetrics@1.2.0
    allow: ^py-flake8@2.6.2
py-msal-extensions@1.0.0
    allow: ^py-portalocker@2.5.1
py-multiecho@0.28
    allow: ^python@3.10.0:3.12.1
py-mxfold2@0.1.1
    added: py-numpy, py-tqdm
     drop: ^py-torch@2.0.0:main
    allow: ^py-torch@1.4.1:1.6.0
py-nanobind
  removed: master
py-nanobind@1.2.0:1.8.0
    allow: ^python@3.7.0:3.7.17
py-nbclassic@0.4.8:1.0.0
    allow: ^py-jupyter-server@1.9.0:1.13.5
py-nbclient@0.5.5
    added: py-async-generator
py-nbclient@0.8.0
     drop: ^python@3.7.0:3.7.17
py-nbconvert@4.0.0:4.1.0
  removed: py-traitlets, py-nbformat, py-jinja2, py-tornado, py-pygments, py-entrypoints, py-jupyter-core, py-mistune
py-nbconvert@4.2.0
    allow: ^py-mistune@0.5:0.5.1,0.7:0.8,2.0.2:2.0.5
    allow: ^py-tornado@6.0.3:6.3.3
py-nbconvert@5.5.0
  removed: py-setuptools
    allow: ^py-mistune@2.0.2:2.0.5
py-nbconvert@5.6.0:6.5.1
  removed: py-setuptools
py-nbconvert@7.14.1
     drop: ^python@3.7.0:3.7.17
py-nbdime@3.1.1
  removed: py-jupyterlab
py-nbformat@4.0.0:4.0.1
  removed: py-jsonschema, py-ipython-genutils, py-jupyter-core, py-traitlets
py-nbformat@4.1.0
    allow: ^py-traitlets@4.0.0
py-nbsphinx@0.8.0:0.8.8
  removed: pandoc
py-ndg-httpsclient@0.5.0:0.5.1
  removed: py-pyopenssl, py-pyasn1
py-neovim-remote@2.4.0
    added: py-setuptools
py-netpyne
  removed: master
py-netpyne@1.0.3.1
     drop: ^py-matplotlib-scalebar@master
py-nibabel@2.4.1:3.2.1
  removed: py-setuptools
py-nibabel@3.2.2:4.0.2
    allow: ^py-setuptools@20.6.7:25.2.0
py-nitransforms
  removed: 20.0.0rc5
    added: 20.0.0-rc5
py-nosexcover@1.0.11
  removed: py-coverage, py-nose
py-notebook@4.0.2:4.0.6
  removed: py-jupyter-client, py-terminado, py-traitlets, py-ipykernel, py-nbformat, py-jinja2, py-ipython-genutils, py-nbconvert, py-jupyter-core, py-tornado
py-notebook@5.7.0:5.7.8
    added: py-send2trash, py-pyzmq
    allow: ^py-nbconvert@4.0.0:4.2.0
py-notebook@6.0.0:6.4.5
    allow: ^py-nbconvert@4.0.0:4.2.0
py-nptyping@1.0.1:1.4.1
  removed: py-typing-extensions
    allow: ^py-numpy@1.17.3:1.21.4,1.21.6:main
py-nptyping@2.4.1
  removed: py-typish
py-ocnn@2.2.0
    allow: ^py-torch@1.4.1:1.5.1
py-omegaconf
  removed: 2.1.0
py-onnxmltools@1.10.0:1.11.1
  removed: py-onnxruntime
py-openai@0.27.8
    allow: ^python@3.7.0
py-opendatalab@0.0.9
    added: py-pywin32
py-openpmd-viewer@1.2.0
  removed: py-numba, py-jupyter, openpmd-api
py-openpmd-viewer@1.3.0:1.4.0
  removed: py-jupyter, openpmd-api
    added: py-ipympl
py-os-service-types@1.7.0
    added: py-pbr
py-oslo-serialization@4.1.0
    added: py-pbr
py-oslo-utils@4.9.2
    added: py-pbr
py-panedr@0.7.1
    allow: ^py-pyedr@0.7.2
py-panedr@0.7.2
    allow: ^py-pyedr@0.7.1
py-panel@0.14.4
    allow: ^py-bokeh@2.4.0:2.4.1
py-paramiko@2.1.2
  removed: py-invoke
py-parsl
  removed: 2023.08.21
    added: 2023.8.21
py-particle
  removed: master
py-particle@0.11.0:0.15.1
    added: py-importlib-resources
     drop: ^py-hepunits@master
py-particle@0.16.0:0.20.0
    added: py-deprecated
     drop: ^py-hepunits@master
py-particle@0.20.1:0.23.0
     drop: ^py-hepunits@master
py-pastml@1.9.38
    allow: ^python@3.7.0:3.7.17
py-pathlib2@2.1.0
  removed: py-six
py-pathy@0.10.1
    added: py-dataclasses
py-patsy@0.4.1:0.5.3
  removed: py-scipy
py-pbr@1.8.1:5.10.0
  removed: py-setuptools
py-pennylane
  removed: master
py-pennylane@0.29.1
  removed: py-pip
    allow: ^py-pennylane-lightning@0.30.0:master
    allow: ^py-scipy@1.11.0:main
    allow: ^python@3.7.0:3.7.17
py-pennylane@0.30.0
  removed: py-pip
    allow: ^py-scipy@1.11.0:main
    allow: ^python@3.7.0:3.7.17
py-pennylane@0.31.0
  removed: py-pip
     drop: ^py-scipy@1.10.1
    allow: ^python@3.7.0:3.7.17
py-pennylane@0.32.0
  removed: py-pip
    allow: ^python@3.7.0:3.8.18
py-pep8@1.7.1
  removed: py-setuptools
py-phylophlan@3.0.2:3.0.3
  removed: blast-plus, iqtree2, muscle, mash, raxml, diamond, trimal, mafft, fasttree
    allow: ^py-biopython@1.65:1.70
    allow: ^py-dendropy@3.12.0:4.3.0
    allow: ^py-matplotlib@2.0.0:3.0.2
    allow: ^py-seaborn@0.7.1
py-picmistandard
  removed: master
py-picmistandard@0.0.18:0.26.0
    allow: ^py-scipy@main
py-pika@0.13.0:1.1.0
  removed: py-gevent
py-planet@1.4.6
  removed: py-click, py-requests-futures, py-requests
py-planetary-computer@0.4.9
    added: py-pytz
py-plotext
  removed: master
py-plotnine@0.7.0:0.7.1
     drop: ^py-mizani@0.6.0
py-pluggy@0.6.0:0.9.0,1.4.0
  removed: py-importlib-metadata
py-poetry-plugin-export@1.0.7:1.6.0
    added: py-poetry
py-poetry@1.1.12:1.1.13
    added: py-cachecontrol
py-poetry@1.2.0
    added: py-cachecontrol
     drop: ^py-cleo@1.0.0a5
    allow: ^py-cleo@1.0.0-alpha5
     drop: ^py-importlib-metadata@5.1.0:7.0.1
    allow: ^py-virtualenv@1.11.6:16.7.6
py-poetry@1.2.1
     drop: ^py-cleo@1.0.0a5
    allow: ^py-cleo@1.0.0-alpha5
     drop: ^py-importlib-metadata@5.1.0:7.0.1
py-poetry@1.6.1
  removed: py-backports-cached-property
    allow: ^py-xattr@0.10.1
py-portalocker@2.5.1
    added: py-pywin32
py-pre-commit@1.20.0:2.17.0
    added: py-importlib-resources
py-progressbar2@3.39.3:3.55.0
  removed: py-python-utils, py-six
py-prompt-toolkit@1.0.9
  removed: py-wcwidth, py-six
    allow: ^python@3.10.0:3.12.1
py-prompt-toolkit@1.0.16:2.0.10
  removed: py-wcwidth, py-six
py-properscoring@0.1
  removed: py-scipy, py-numpy
py-prov@1.5.1
  removed: graphviz
    added: py-six
    allow: ^py-lxml@2.3
    allow: ^py-networkx@1.10:1.11
    allow: ^py-python-dateutil@1.5
py-prov@2.0.0
  removed: graphviz
py-psmon@1.1.1
     drop: ^py-loguru@0.3.0
py-pulsar-galaxy-lib@0.14.16
  removed: py-galaxy-tool-util, py-galaxy-util, py-galaxy-job-metrics, py-galaxy-objectstore
py-pure-eval
  removed: master
py-pyaestro
  removed: 0.0.1a2,main
    added: 0.0.1-alpha2
py-pyani@0.2.6:0.2.7
  removed: blast-plus, mummer
py-pycodestyle@2.0.0:2.10.0
  removed: py-setuptools
py-pycodestyle@2.11.0
  removed: py-setuptools
     drop: ^python@3.7.0:3.7.17
py-pydicom@2.1.2:2.3.0
  removed: py-setuptools, py-numpy
py-pydicom@2.4.1:2.4.3
  removed: py-numpy
py-pydmd/isuue@0.3
    added: py-sphinx, py-sphinx-rtd-theme
py-pydmd@0.3
    added: py-sphinx, py-sphinx-rtd-theme
py-pydub@0.25.1
  removed: ffmpeg
py-pyepsg@0.3.2
  removed: py-requests
py-pyflakes@0.9.0:3.0.1
  removed: py-setuptools
py-pyflakes@3.1.0
  removed: py-setuptools
     drop: ^python@3.7.0:3.7.17
py-pygithub@1.55
    allow: ^py-pyjwt@2.1.0
py-pyglet@1.2.1
  removed: py-future
py-pyglet@2.0.9:2.0.10
  removed: pulseaudio
    allow: ^python@3.7.0:3.7.17
py-pygments-pytest@1.2.0
    added: py-pygments
py-pygments@2.0.1:2.16.1
  removed: py-setuptools
py-pyh5md@1.0.0
  removed: py-h5py, py-numpy
py-pykwalify@1.6.1
  removed: py-docopt, py-ruamel-yaml, py-pyyaml, py-python-dateutil
py-pykwalify@1.7.0
  removed: py-ruamel-yaml
    added: py-pyyaml
py-pylab-sdk@1.3.2
  removed: py-requests
py-pylint@2.8.2:2.13.5
    allow: ^py-colorama@0.3.7:0.4.4
py-pylint@2.14.4:2.16.2
    allow: ^python@3.7.0:3.7.1
py-pylops
  removed: master
py-pylops@1.11.1:1.12.0
     drop: ^py-spgl1@master
py-pympler@1.0.1
    allow: ^python@3.11.0:3.12.1
py-pyopenssl
    added: 20.0.1
py-pyperf
  removed: 1.6.0,1.5.1
py-pyqtgraph@0.13.3
  removed: py-pyqt6
py-pyquaternion@0.9.5
  removed: py-setuptools
py-pyshacl@0.17.2
     drop: ^py-owlrl@6.0.2
py-pyshacl@0.25.0
    allow: ^py-importlib-metadata@6.6.0
    allow: ^python@3.8.0
py-pysurfer@0.11.0
  removed: py-imageio
py-pytaridx
  removed: master
py-pytest-asyncio@0.18.3
    added: py-typing-extensions
py-pytest-asyncio@0.23.5
     drop: ^python@3.7.0:3.7.17
py-pytest-check-links@0.3.4
    added: py-nbformat, py-docutils, py-pytest, py-six, py-nbconvert, py-html5lib
py-pytest-cov@2.3.1
    allow: ^py-coverage@4.0a6:4.3.4
    allow: ^py-pytest@3.0.2:3.5.1
py-pytest-flake8@0.8.1
  removed: py-pytest, py-flake8
py-pytest-html@3.2.0
    added: py-py
py-pytest-isort@0.3.1
  removed: py-pytest, py-isort
py-pytest-mock@1.2
  removed: py-pytest
py-pytest-mpi@0.6
    allow: ^py-pytest@3.0.2:3.0.7
py-pytest-runner@2.11.1:6.0.0
  removed: py-setuptools
py-pytest@3.0.2:3.0.7
  removed: py-py, py-setuptools, py-colorama, py-pluggy
    allow: ^python@3.12.0:3.12.1
py-pytest@3.5.1:3.7.2
    allow: ^python@3.12.0:3.12.1
py-pytest@4.3.0:4.4.0
    allow: ^py-setuptools@20.6.7:39.2.0
    allow: ^python@3.12.0:3.12.1
py-pytest@4.6.2:6.1.1,6.2.5,7.2.1
  removed: py-setuptools
    allow: ^python@3.12.0:3.12.1
py-pytest@6.2.1:6.2.4
  removed: py-setuptools
     drop: ^py-pluggy@1.0.0:1.4.0
    allow: ^python@3.12.0:3.12.1
py-pytest@7.1.3
  removed: py-exceptiongroup, py-setuptools
    allow: ^python@3.12.0:3.12.1
py-pytest@7.3.2:7.4.4
  removed: py-setuptools
py-pytest@8.0.0
  removed: py-setuptools, py-importlib-metadata
py-python-box@5.3.0
  removed: py-toml, py-ruamel-yaml, py-msgpack
py-python-certifi-win32@1.6
    added: py-setuptools-scm
py-python-daemon@2.0.5
  removed: py-lockfile, py-setuptools
py-python-daemon@2.3.0:2.3.1
    added: py-docutils
py-python-dotenv@1.0.1
     drop: ^python@3.7.0:3.7.17
py-python-engineio@2.0.2
  removed: py-six
py-python-keystoneclient@4.2.0
    added: py-pbr
py-python-lsp-jsonrpc@1.1.2
     drop: ^python@3.7.0:3.7.17
py-python-lsp-server@1.6.0:1.7.1
    allow: ^py-setuptools@39.0.1:59.4.0
py-python-magic@0.4.15:0.4.24
  removed: file
py-python-picard@0.6
  removed: py-scikit-learn, py-scipy, py-matplotlib
py-python-socketio@1.8.4
  removed: py-python-engineio, py-six, py-eventlet
py-python-subunit@1.3.0
  removed: py-extras, py-testtools
py-python-utils@2.3.0:2.4.0
  removed: py-six
py-python-xmp-toolkit@2.0.1
  removed: exempi
py-python3-openid@3.2.0
    added: py-defusedxml
py-pythonqwt@0.5.5
  removed: py-sip, py-numpy, py-pyqt4
py-pytorch-lightning
  removed: 1.7.2,1.6.3,1.8.0,1.6.0,1.8.2,1.7.3,1.6.4,1.7.4,1.7.6,1.8.1,1.6.1,1.7.0,1.6.2,1.6.5,1.7.5,1.8.3,1.7.7,1.7.1
py-pytorch-lightning@1.2.10
    added: py-onnxruntime, py-omegaconf, py-torchtext, py-horovod, py-hydra-core, py-matplotlib
py-pytorch-lightning@1.3.8
    added: py-onnxruntime, py-omegaconf, py-jsonargparse, py-torchtext, py-horovod, py-hydra-core, py-matplotlib, py-pillow
py-pytorch-lightning@1.4.0:1.4.1
    added: py-onnxruntime, py-omegaconf, py-onnx, py-gcsfs, py-jsonargparse, py-torchtext, py-horovod, py-hydra-core, py-matplotlib
py-pytorch-lightning@1.5.3
    added: py-omegaconf, py-rich, py-gcsfs, py-jsonargparse, py-torchtext, py-horovod, py-hydra-core, py-matplotlib
py-pytorch-lightning@1.8.4:1.8.6
    added: py-omegaconf, py-rich, py-jsonargparse, py-hydra-core, py-matplotlib
py-pytorch-lightning@1.9.0:2.0.0
    added: py-omegaconf, py-rich, py-jsonargparse, py-tensorboardx, py-hydra-core, py-matplotlib
py-pytorch-sphinx-theme
  removed: master
    added: 0.0.19
py-pyutilib
  removed: 5.3.4,5.6.1
py-pyutilib@5.3.3:5.3.5,5.4.1,5.5.1,5.6.2
  removed: py-six, py-nose
py-pyvistaqt
  removed: 0.5.0
    added: 0.11.0
py-qiskit-nature@0.2.2
    added: py-setuptools, py-dataclasses
py-qtawesome@0.3.3:0.4.1
  removed: py-qtpy, py-six
py-qtconsole@4.2.1:4.5.1
  removed: py-sphinx, py-pyqt5
py-qtconsole@5.2.0
  removed: py-sphinx
py-quaternionic@1.0.1
    added: py-black
     drop: ^py-numpy@main
py-radiant-mlhub@0.3.0:0.4.0
  removed: py-typing-extensions
py-rarfile@4.0:4.1
  removed: unrar
py-rdflib@7.0.0
  removed: py-importlib-metadata
    allow: ^python@3.8.0
py-rdt
    added: 0.6.4,1.10.0
py-rdt@0.6.1
     drop: ^py-numpy@1.20.0:1.26.4
    allow: ^py-scipy@main
     drop: ^python@3.10.0:3.12.1
py-readme-renderer@16.0
    allow: ^py-bleach@1.5.0
py-recommonmark@0.6.0
    allow: ^py-sphinx@1.4
py-reindent@3.5.1
    added: py-setuptools
py-repoze-lru@0.7
    added: py-sphinx
py-requests-mock@1.7.0
    added: py-six, py-requests
py-requests@2.3.0:2.11.1
  removed: py-pysocks
py-requests@2.18.4
     drop: ^py-idna@2.8
     drop: ^py-urllib3@1.24.3:1.25.11
py-requests@2.21.0
     drop: ^py-urllib3@1.25.3:1.25.11
py-requests@2.26.0:2.28.0
     drop: ^py-charset-normalizer@2.1.1
py-rich@9.4.0:12.5.1
    added: py-dataclasses
py-robotframework@3.2.1:3.2.2
  removed: py-setuptools
py-rosdep@0.20.0
  removed: py-setuptools
py-rospkg@1.2.9
  removed: py-setuptools
py-rq@1.2.1:1.4.3
    allow: ^py-redis@3.3.8
    allow: ^python@3.9.0:3.12.1
py-rq@1.5.0:1.5.2
    allow: ^python@3.9.0:3.12.1
py-rsa@3.4.2:4.0
  removed: py-pyasn1, py-setuptools
py-rsa@4.7.2
  removed: py-setuptools
py-rseqc@3.0.1:5.0.1
  removed: r
py-s3fs@2024.2.0
     drop: ^python@3.7.0:3.7.17
py-s3transfer@0.2.1
  removed: py-botocore
py-s3transfer@0.9.0:0.10.0
     drop: ^python@3.7.0:3.7.17
py-sacrebleu
    added: 2.4.1
py-sanic@20.6.3
  removed: py-brotlipy, py-urllib3, py-h2, py-chardet, py-hstspreload
py-scikit-build-core@0.2.0:0.6.1
  removed: cmake
py-scikit-build@0.15.0
    allow: ^py-setuptools@39.0.1:41.4.0
py-scp@0.13.2
  removed: py-paramiko
py-sdmetrics@0.4.1
    allow: ^py-copulas@0.6.1
     drop: ^py-numpy@1.20.0:1.26.4
    allow: ^py-pandas@1.1.5:1.5.3
    allow: ^py-rdt@0.6.4
    allow: ^py-scikit-learn@master
    allow: ^py-scipy@main
     drop: ^python@3.10.0:3.12.1
py-sdv@0.13.1
    allow: ^py-copulas@0.6.1
    allow: ^py-ctgan@0.5.2
     drop: ^py-numpy@1.20.0:1.26.4
    allow: ^py-pandas@1.1.5:1.5.3
    allow: ^py-rdt@0.6.4
     drop: ^python@3.10.0:3.12.1
py-sdv@0.14.0
     drop: ^py-copulas@0.6.0
    allow: ^py-copulas@0.6.1
     drop: ^py-ctgan@0.5.0
    allow: ^py-ctgan@0.5.2
     drop: ^py-numpy@1.20.0:1.26.4
    allow: ^py-pandas@1.1.5:1.5.3
     drop: ^py-rdt@0.6.1
    allow: ^py-rdt@0.6.4
     drop: ^python@3.10.0:3.12.1
py-segmentation-models-pytorch@0.3.0:0.3.3
    added: py-pillow
py-selenium@3.141.0
  removed: py-urllib3
py-sentry-sdk@0.17.6
  removed: py-asttokens, py-quart, py-executing, py-pure-eval, py-httpx
     drop: ^py-celery@4.3.0,4.4.7,5.0.0
py-sentry-sdk@1.5.5
  removed: py-pure-eval, py-executing, py-asttokens
     drop: ^py-celery@4.3.0,4.4.7,5.0.0
py-setupmeta@3.3.0:3.4.0
  removed: py-setuptools
py-setuptools-cpp@0.1.0
  removed: py-certifi, py-pathspec, py-mccabe, py-jinja2, py-coverage, py-atomicwrites, py-attrs, py-appdirs, py-chardet, py-entrypoints, py-mypy-extensions, py-typed-ast, py-flake8, py-more-itertools, py-isort, py-tornado, py-markupsafe, py-mypy, py-click, py-importlib-metadata, py-markdown, py-packaging, py-colorama, py-black, py-pluggy
    added: py-pybind11, py-setuptools
py-setuptools-git-versioning@1.13.3
  removed: git
py-setuptools-git@1.2
  removed: git
py-setuptools-scm-git-archive@1.0:1.4
  removed: py-setuptools-scm
py-setuptools-scm@1.15.6:3.5.0
  removed: py-toml, git, py-setuptools
py-setuptools-scm@4.1.2
  removed: git
    allow: ^py-setuptools@20.6.7:25.2.0
py-setuptools-scm@5.0.2
  removed: git
    allow: ^py-setuptools@20.6.7:41.4.0
py-setuptools-scm@6.0.1
  removed: git
py-setuptools-scm@6.3.2:7.1.0
  removed: git
    allow: ^py-setuptools@43.0.0:44.1.1
py-sierrapy@0.3.0
  removed: py-idna, py-click, py-urllib3, py-tqdm, py-gql, py-voluptuous, py-graphql-core, py-promise, py-chardet, py-rx, py-six, py-requests
py-skl2onnx@1.10.3
    allow: ^py-scikit-learn@1.1.2:master
py-skl2onnx@1.12
  removed: py-packaging
    allow: ^py-scikit-learn@1.1.2:master
py-smac@1.1.1
    allow: ^py-scikit-learn@0.22
py-snakemake-interface-common@1.17.1
     drop: ^python@3.7.0:3.7.17
py-snakemake-interface-report-plugins@1.0.0
     drop: ^python@3.7.0:3.10.13
py-snakemake-storage-plugin-gcs@0.1.3
  removed: py-snakemake-interface-common, py-snakemake-interface-storage-plugins
py-soyclustering@0.2.0
  removed: py-scikit-learn, py-scipy, py-matplotlib
py-spatialite@0.0.3
  removed: libspatialite
py-spgl1
  removed: master
py-sphinx-argparse@0.3.1
    allow: ^py-sphinx@1.4
py-sphinx-basic-ng
  removed: 1.0.0b2
    added: 1.0.0-beta2
py-sphinx-multiversion@0.2.1:0.2.4
     drop: ^py-sphinx@1.3.1,1.4.5:1.8.5
py-sphinx-removed-in@0.2.1
    allow: ^py-sphinx@1.4
py-sphinx-rtd-theme@0.4.3:0.5.2
    allow: ^py-sphinx@1.4,7.0.0:7.2.6
py-sphinx-rtd-theme@1.0.0
    allow: ^py-sphinx@7.0.0:7.2.6
py-sphinx-tabs@3.4.1:3.4.4
    allow: ^py-sphinx@1.4
py-sphinx-theme-builder
  removed: 0.2.0b2
    added: 0.2.0-beta2
py-sphinx
    added: 1.4
py-sphinx@1.3.1
  removed: py-snowballstemmer, py-alabaster, py-docutils, py-jinja2, py-sphinx-rtd-theme, py-pygments, py-setuptools, py-colorama, py-six, py-babel
py-sphinx@1.4.5
  removed: py-snowballstemmer, py-alabaster, py-imagesize, py-docutils, py-jinja2, py-pygments, py-setuptools, py-colorama, py-six, py-babel
py-sphinx@1.5.5
  removed: py-snowballstemmer, py-alabaster, py-imagesize, py-docutils, py-jinja2, py-pygments, py-setuptools, py-colorama, py-six, py-babel, py-requests
py-sphinx@1.6.1:1.6.3
  removed: py-snowballstemmer, py-alabaster, py-sphinxcontrib-websupport, py-imagesize, py-docutils, py-jinja2, py-pygments, py-setuptools, py-colorama, py-six, py-babel, py-requests
py-sphinx@1.7.4
  removed: py-snowballstemmer, py-alabaster, py-sphinxcontrib-websupport, py-imagesize, py-docutils, py-jinja2, py-packaging, py-pygments, py-setuptools, py-colorama, py-six, py-babel, py-requests
py-sphinx@1.8.2:1.8.5
    allow: ^py-docutils@0.17:0.20.1
    allow: ^py-jinja2@3.0.1:3.1.2
    allow: ^py-requests@2.3.0
py-sphinx@2.2.0:3.4.1
    allow: ^py-docutils@0.17:0.20.1
    allow: ^py-jinja2@3.0.1:3.1.2
py-sphinx@3.5.4
    allow: ^py-jinja2@3.0.1:3.1.2
py-sphinx@4.0.0:4.0.1
    added: py-markupsafe
py-sphinxcontrib-bibtex@0.3.5
  removed: py-sphinx, py-oset, py-latexcodec, py-six, py-pybtex-docutils, py-pybtex
py-sphinxcontrib-bibtex@1.0.0
     drop: ^py-sphinx@1.3.1,1.4.5:1.8.5
py-sphinxcontrib-bibtex@2.2.0:2.5.0
    added: py-dataclasses
py-sphinxcontrib-jquery@2.0.0
  removed: py-sphinx
    allow: ^py-setuptools@20.6.7:64.0.0
py-sphinxcontrib-mermaid@0.4.0:0.7.1
  removed: py-sphinx
py-sphinxcontrib-moderncmakedomain@3.25.0:3.26.4
    allow: ^py-sphinx@1.4
py-sphinxcontrib-programoutput@0.10
    allow: ^py-sphinx@1.4:1.6.3
py-sphinxcontrib-serializinghtml@1.1.9
    added: py-sphinx
py-sphobjinv@2.3.1
  removed: py-setuptools
py-sqlalchemy-migrate@0.13.0
    added: py-pbr
py-stack-data@0.2.0:0.6.2
     drop: ^py-pure-eval@master
py-stevedore@1.28.0
    allow: ^py-pbr@3.1.1:5.10.0
py-stevedore@3.5.0
    added: py-importlib-metadata
py-stomp-py@8.0.0
    added: py-pyopenssl
py-subrosa@0.1.0
  removed: py-gf256
py-superqt@0.6.1
     drop: ^py-typing-extensions@3.10.0.0
py-svgwrite@1.1.12
  removed: py-pyparsing
py-sysrsync@1.1.1
  removed: rsync
py-tatsu@4.4.0
  removed: py-regex
py-tensorboardx@1.8
    allow: ^py-protobuf@3.3.0:3.7.1,3.20.2:4.24.3
py-tensorboardx@1.9:2.1
    allow: ^py-protobuf@3.20.2:4.24.3
py-tensorflow-datasets@4.4.0
    added: py-dataclasses
py-tensorflow-probability
  removed: 0.18.0,0.22.0,0.12.1,0.23.0,0.20.0,0.22.1,0.19.0
    added: 0.20.1
py-tensorflow-probability@0.21.0
  removed: py-jax, py-tensorflow
    added: py-typing-extensions
py-textual@0.47.1
  removed: py-mdit-py-plugins
py-timm@0.4.12:0.5.4
  removed: py-numpy
    allow: ^python@3.11.0:3.12.1
py-timm@0.6.12
    allow: ^python@3.11.0:3.12.1
py-timm@0.9.0
    allow: ^python@3.7.0:3.7.17
py-tinydb@4.5.2
     drop: ^py-typing-extensions@4.1.1:4.8.0
py-tldextract@5.1.1
     drop: ^python@3.7.0:3.7.17
py-toposort@1.8:1.10
    allow: ^python@3.7.0:3.7.17
py-torchmeta@1.7.0
    added: py-pillow
py-torchmetrics@1.0.2:1.3.2
     drop: ^python@3.7.0:3.7.17
py-torchsummary@1.5.1
  removed: py-torch, py-numpy
py-tox@3.14.2
  removed: py-setuptools
    added: py-colorama
     drop: ^py-importlib-metadata@2.0.0:7.0.1
py-tpot@0.11.5:0.11.7
    allow: ^py-scikit-learn@0.22
py-tqdm@4.8.4:4.46.0
  removed: py-requests, py-setuptools, py-colorama, py-ipywidgets
py-tqdm@4.56.2:4.58.0
  removed: py-setuptools, py-colorama, py-ipywidgets
py-tqdm@4.59.0
  removed: py-setuptools, py-colorama
py-tqdm@4.62.3,4.65.0:4.66.1
  removed: py-setuptools
py-tqdm@4.64.0:4.64.1
  removed: py-setuptools
    added: py-importlib-resources
py-traceback2@1.4.0
  removed: py-linecache2, py-six
py-traitlets@4.0.0
  removed: py-ipython-genutils, py-decorator, py-six
py-traitlets@4.1.0:4.2.2
  removed: py-six
py-transformers@2.8.0
    added: py-dataclasses
py-transformers@4.6.1
    added: py-importlib-metadata, py-dataclasses
py-transformers@4.24.0
    added: py-importlib-metadata
py-transformers@4.31.0:4.35.2
     drop: ^python@3.7.0:3.7.17
py-trx-python@0.2.9
    allow: ^py-setuptools-scm@1.15.6:3.3.3
     drop: ^python@3.7.0:3.7.17
py-twine@2.0.0
    added: py-setuptools
py-typeguard@3.0.2
    allow: ^python@3.7.0:3.7.3
py-uncertainties@3.1.4:3.1.7
    added: py-sphinx
py-unidecode
  removed: 0.04.21
    added: 0.4.21
py-unyt
  removed: main
py-uproot@4.0.6:4.0.7
  removed: zstd, lz4, py-setuptools, xrootd, xxhash
py-uproot@4.0.8:4.0.11,5.0.0:5.0.5
  removed: zstd, lz4, xxhash, xrootd
py-uproot@4.1.8:4.3.7
  removed: zstd, lz4, xxhash, xrootd
    allow: ^py-setuptools@20.6.7:41.4.0
py-urllib3@1.14
  removed: py-brotlipy, py-idna, py-pyopenssl, py-pysocks, py-cryptography
py-urllib3@1.20:1.21.1
  removed: py-brotlipy, py-idna, py-cryptography, py-pyopenssl
py-urllib3@1.24.3
  removed: py-brotlipy
    added: py-ipaddress
    allow: ^py-pyopenssl@20.0.1
py-urllib3@1.25.3:2.0.7
    allow: ^py-pyopenssl@20.0.1
py-urllib3@2.1.0
  removed: py-pyopenssl, py-certifi, py-idna, py-urllib3-secure-extra, py-cryptography
py-uvw
  removed: master
py-uvw@0.0.7
  removed: py-mpi4py
py-versioneer@0.18
  removed: py-tomli
py-virtualenv@1.11.6:16.7.6
  removed: py-setuptools
py-virtualenv@20.10.0:20.17.1
  removed: py-setuptools
    added: py-importlib-resources
py-websocket-client@0.48.0:0.57.0
  removed: py-six
py-werkzeug@0.11.11:0.12.2
    allow: ^python@3.10.0:3.12.1
py-werkzeug@2.0.2
    added: py-dataclasses
py-widgetsnbextension@1.2.6
  removed: py-notebook
py-wsproto@1.0.0
    added: py-dataclasses
py-ww@0.2.1
  removed: py-chardet, py-formatizer, py-future, py-six
py-xarray-tensorstore@0.1.1
     drop: ^python@3.7.0:3.9.18
py-xenv
  removed: develop
py-xenv@1.0.0
  removed: py-setuptools
py-xesmf@0.8.4
     drop: ^python@3.7.0:3.7.17
py-yaspin@2.1.0
    added: py-dataclasses
py-youtube-dl@2020.3.24:2021.12.17
  removed: py-setuptools, ffmpeg
py-ytopt
  removed: 0.0.2,0.0.4,0.0.3
py-ytopt@0.0.1
  removed: py-ray, py-dh-scikit-optimize, py-ytopt-autotune, py-configspace, py-deap
    added: py-gym, py-scikit-optimize, py-tensorflow, py-keras
    allow: ^py-scikit-learn@0.22:0.23.0,0.23.2:master
py-zfit-interface@0.0.3
    allow: ^python@3.7.0:3.7.17

```

haampie commented:
@spackbot run pipeline
